### PR TITLE
Provide API for VCS mappings

### PR DESCRIPTION
**Pull request by big-guy**

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/vcs/SourceControl.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+package org.gradle.api.vcs;
+
+import org.gradle.api.Action;
+import org.gradle.api.Incubating;
+
+/**
+ * In settings.gradle:
+ * <pre>
+ * sourceControl {
+ *    repositories {
```
Comment (adammurdoch) [inline subprojects/core-api/src/main/java/org/gradle/api/vcs/SourceControl.java]: I'd go with something more mapping-like, where "mapping" means a function that given a component selector of some kind returns either the VCS URL for that component or "don't know".

In other words, this should look more like the other kinds of component meta-data rules we have, where I can either provide the function (as an `Action` that takes some details) or a hard-coded mapping ("map this component id to this VCS URL").

For example:

```
components {
   withModule("group:name") { 
       // If you need something with maven group and artifact id (group, name), then you can find it here:
       vcsUrl = "https://..."
   }
   all { details ->
      if (details.group == "my-group") {
         vcsUrl = "https://git.my.org/${details.name}"
      }
   }
}
```

File: subprojects/core/src/integTest/groovy/org/gradle/api/vcs/internal/SourceControlIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+ */
+
+package org.gradle.api.vcs.internal
+
+import org.gradle.api.vcs.VcsRepository
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+class SourceControlIntegrationTest extends AbstractIntegrationSpec {
+    def setup() {
+        settingsFile << """
```
Comment (adammurdoch) [inline subprojects/core/src/integTest/groovy/org/gradle/api/vcs/internal/SourceControlIntegrationTest.groovy]: I shouldn't have to do any of this stuff as a user of this feature

File: subprojects/core-api/src/main/java/org/gradle/api/vcs/internal/MavenVcsMapping.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.api.vcs.internal;
+
+import org.gradle.api.vcs.VcsMapping;
+
+public interface MavenVcsMapping extends VcsMapping {
+    String getGroup();
+    String getModule();
```
Comment (adammurdoch) [inline subprojects/core-api/src/main/java/org/gradle/api/vcs/internal/MavenVcsMapping.java]: we should use `ComponentSelector` to represent identifiers in a namespace (the maven namespace in this instance).

File: subprojects/core-api/src/main/java/org/gradle/vcs/VcsMappings.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * In settings.gradle:
+ * <pre>
+ * vcsMappings {
+ *   withModule("group:name") {
+ *     from vcs(GitVcs) {
+ *         url = "..."
+ *     }
+ *   }
+ *   all { details ->
```
Comment (eljobe) [inline subprojects/core-api/src/main/java/org/gradle/vcs/VcsMappings.java]: I don't really understand the usecase this supports.

File: subprojects/core-api/src/main/java/org/gradle/vcs/VcsMapping.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import org.gradle.api.Incubating;
+import org.gradle.api.artifacts.component.ComponentSelector;
+
+/**
+ * A dependency mapping provided by a VCS repository.
+ *
+ * @since 4.3
+ */
+@Incubating
+public interface VcsMapping {
```
Comment (eljobe) [inline subprojects/core-api/src/main/java/org/gradle/vcs/VcsMapping.java]: This has drifted a little bit from the original design. I'm not sure now how a build master would provide an implementation for VcsMapping that would implement some pattern her company uses for mapping from a component selector to a vcs repository. Let's discuss.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyServices.java [hunk 6]
```diff
@@ -38,21 +38,25 @@ public void registerGradleUserHomeServices(ServiceRegistration registration) {
...
-                return new DefaultCacheLockingManager(cacheRepository, artifactCacheMetaData);
-            }
-
-            TransformedFileCache createTransformedFileCache(ArtifactCacheMetaData artifactCacheMetaData, CacheRepository cacheRepository, InMemoryCacheDecoratorFactory cacheDecoratorFactory, FileSystemSnapshotter fileSystemSnapshotter, ListenerManager listenerManager) {
-                DefaultTransformedFileCache transformedFileCache = new DefaultTransformedFileCache(artifactCacheMetaData, cacheRepository, cacheDecoratorFactory, fileSystemSnapshotter);
-                listenerManager.addListener(transformedFileCache);
-                return transformedFileCache;
-            }
-        });
+        registration.addProvider(new DependencyManagementBuildSessionServices());
```
Comment (eljobe) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyServices.java]: This diff looks unrelated to your changes?

File: subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVcsMapping.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public ModuleVersionSelector getOldRequested() {
+        return oldRequested;
+    }
+
+    @Override
+    public void from(VersionControlSpec versionControlSpec) {
+        Preconditions.checkNotNull(versionControlSpec, "VCS repository cannot be null");
```
Comment (eljobe) [inline subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVcsMapping.java]: "VCS specification cannot be null."

File: subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVcsMapping.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+import org.gradle.vcs.VersionControlSpec;
+
+public class DefaultVcsMapping implements VcsMappingInternal {
+    private final ComponentSelector requested;
+    private final ModuleVersionSelector oldRequested;
+    private VersionControlSpec versionControlSpec;
+
+    public DefaultVcsMapping(ComponentSelector requested, ModuleVersionSelector oldRequested) {
+        this.requested = requested;
+        this.oldRequested = oldRequested;
```
Comment (eljobe) [inline subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVcsMapping.java]: I don't really understand just from the variable names and types what `requested` and `oldRequested` are supposed to be representing.

File: subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVersionControlSystemFactory.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import org.gradle.vcs.VersionControlSystem;
+import org.gradle.vcs.git.internal.GitVersionControlSystem;
+
+public class DefaultVersionControlSystemFactory implements VersionControlSystemFactory {
+    @Override
+    public VersionControlSystem create(VersionControlSpec spec) {
+        // TODO: Register these mappings somewhere
+        if (spec instanceof DirectoryRepository) {
+            return new SimpleVersionControlSystem();
+        }
```
Comment (eljobe) [inline subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVersionControlSystemFactory.java]: We should explicitly check for `GitVersionControlSpec` as well. Let the last line return null or throw an appropriate runtime exception.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyServices.java (hunk 6 above)
Comment (big-guy) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DependencyServices.java]: At one point, I had added something to these services, so it was getting ugly to keep the anonymous class around.

File: subprojects/core-api/src/main/java/org/gradle/vcs/VcsMapping.java (hunk 5 above)
Comment (big-guy) [reply subprojects/core-api/src/main/java/org/gradle/vcs/VcsMapping.java]: They wouldn't provide an implementation.  We have the `VcsMappings` DSL to configure a `VcsMapping` instance we create to map a `ComponentSelector` to a `VersionControlSpec`.

File: subprojects/core-api/src/main/java/org/gradle/vcs/VcsMappings.java (hunk 4 above)
Comment (big-guy) [reply subprojects/core-api/src/main/java/org/gradle/vcs/VcsMappings.java]: It would let you do something like:

```
all { details ->
   if (details.requested.group == "com.example") {
      from vcs(GitVcs) {
         url = "http://github.com/com.example/${details.requested.name}"
      }
   }
}
```

But I could see dropping this for now because it opens up a can of worms if you let a single VersionControlSpec map to multiple Maven coordinates.

File: subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVersionControlSystemFactory.java (hunk 9 above)
Comment (big-guy) [reply subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVersionControlSystemFactory.java]: Yep, we should explode if we don't understand the spec type

File: subprojects/core-api/src/main/java/org/gradle/vcs/VcsMappings.java (hunk 4 above)
Comment (eljobe) [reply subprojects/core-api/src/main/java/org/gradle/vcs/VcsMappings.java]: Yeah. That seems like an exceedingly rare use-case. More likely we'd have an implementation of `VcsMapping` that knows how to generate a correct `VersionControlSpec` for some universe of `ComponentSelector`s

File: subprojects/core-api/src/main/java/org/gradle/vcs/VcsMapping.java (hunk 5 above)
Comment (eljobe) [reply subprojects/core-api/src/main/java/org/gradle/vcs/VcsMapping.java]: Yeah, but what if the company has a predictable pattern of how to map from a `ComponentSelector` to a `VersionControlSpec` for their internally-hosted GitHub Enterprise instance. They shouldn't have to add a `VcsMapping` for each repository. They should be able to provide a `VcsMapping` implementation which, given a `ComponentSelector` can return the appropriate `VersionControlSpec`
